### PR TITLE
Reject data writes if statfs suggests fs close to full

### PR DESCRIPTION
**Pull request by irq0**
First step towards ENOSPC handling and SQLite storage reservation.

### CODE REVIEW

File: src/common/options/rgw.yaml.in [hunk 1]
```diff
@@ -3565,3 +3565,18 @@ options:
...
+  default: 10000
+  desc: RGW SFS stats update interval in milliseconds
+  service:
+    - rgw
+- name: rgw_sfs_min_space_left_for_write_ops
+  type: int
+  level: advanced
+  default: 104857600
+  desc: Minimum space left (in bytes) for SFS to attempt atomic or multipart writes.
+    Reserves space for metadata operatoins and SQLite updates.
```
Comment (jecluis) [inline src/common/options/rgw.yaml.in:3580]: *operations

**Review by jecluis — CHANGES_REQUESTED** [review]
will test soon, but looks good minus few things

File: src/rgw/rgw_sal_sfs.cc [hunk 2]
```diff
@@ -502,25 +504,64 @@ void SFStore::maybe_init_store() {
...
-  const std::filesystem::path &data_path
-) : sync_module(),
-    zone(this),
-    data_path(data_path),
-    cctx(c) {
+void SFStore::filesystem_stats_updater_main(
+    std::chrono::milliseconds update_interval) {
+  while (true) {
+    std::unique_lock<std::mutex> lock(filesystem_stats_updater_mutex);
+    ldout(ctx(), 19) << __func__ << ": updating filesystem stats" << dendl;
```
Comment (jecluis) [inline src/rgw/rgw_sal_sfs.cc]: can we please make this `20`? It's unlikely anyone besides whomever wrote this line will ever run rgw at level 19.

File: src/rgw/rgw_sal_sfs.h [hunk 3]
```diff
@@ -111,10 +114,22 @@ class SFStore : public StoreDriver {
   ceph::mutex buckets_map_lock = ceph::make_mutex("buckets_map_lock");
   std::map<std::string, sfs::BucketRef> buckets;
 
+  // Signal shutdown condition to service threads
+  bool shutdown;
+
+  std::thread filesystem_stats_updater;
+  std::condition_variable filesystem_stats_updater_cvar;
+  std::mutex filesystem_stats_updater_mutex;
```
Comment (jecluis) [inline src/rgw/rgw_sal_sfs.h]: why not use `ceph::mutex` here? we sort of lose lockdep capabilities that way, afaiu.

**Review by jecluis — CHANGES_REQUESTED** [review]

File: src/rgw/rgw_sal_sfs.cc (hunk 2 above)
Comment (irq0) [reply src/rgw/rgw_sal_sfs.cc]: :+1:

File: src/rgw/rgw_sal_sfs.h (hunk 3 above)
Comment (irq0) [reply src/rgw/rgw_sal_sfs.h]: :+1:

**Review by jecluis — APPROVED** [review]

**Pull request merged**